Code/src/main/java/nl/utwente/group10/haskell/type/ConstT.java

```diff
@@ -62,6 +67,30 @@ public String toHaskellType() {
     }
 
     @Override
+    public ConstT getFresh() {
+        return new ConstT(this.constructor, this.getFreshArgs());
+    }
+
+    /**
+     * Returns an array of fresh arguments. Selectively calls {@code getFresh} on each argument. When the same Type
+     * instance appears multiple times in the arguments no new type is instantiated. Instead, the fresh type is reused.
+     * @return An array of fresh arguments.
+     */
+    protected Type[] getFreshArgs() {
+        List<Type> fresh = new LinkedList<>();
```

Comment:
Hmm, that's quite possible. However, at the end of the method you call toArray on the whole thing, so... I don't know, I'll leave it up to you.